rclpy/src/rclpy/_rclpy.c

```diff
@@ -2281,9 +2281,14 @@ rclpy_destroy_wait_set(PyObject * Py_UNUSED(self), PyObject * args)
   const rcl_ ## ENTITY_TYPE ## _t ** struct_ptr = wait_set->ENTITY_TYPE ## s; \
   for (idx = 0; idx < idx_max; idx ++) { \
     if (struct_ptr[idx]) { \
-      PyList_Append( \
-        entity_ready_list, \
-        PyLong_FromUnsignedLongLong((uint64_t) & struct_ptr[idx]->impl)); \
+      PyObject* obj = PyLong_FromUnsignedLongLong((uint64_t) & struct_ptr[idx]->impl); \
+      if (obj) { \
+        PyList_Append(entity_ready_list, obj); \
+        Py_DECREF(obj); \
+      } else { \
+        Py_DECREF(entity_ready_list); \
+        return NULL; \
```

Comment:
This `else` changes behavior. Before the code didn't check the value returned by `PyLong_FromUnsignedLongLong` (sure, not great). With the patch it suddenly return `NULL` in that case without any kind of error reporting. I neither think that change is necessary to address the memory leak described in this ticket nor that it should be done like this.

Therefore I would suggest to just remove the `else` block from the patch.

Reply:
I'm pretty sure PyLong_FromUnsignedLongLong is setting exception when it returns NULL. That's how most CPython API functions work - if they fail they set global exception which is "raised" when C function returns and return value will not be used at all.

I feel that by ignoring error case and just skip adding element to list won't inform user of this API that something "bad" happened. Are you sure its ok to so?

Reply:
Here's one place where PyLogn_FromUnsignedLongLong can fail and return NULL: https://github.com/python/cpython/blob/6405feecda6a5d5dd7a4240eb3054a2676ed29b1/Objects/longobject.c#L327
It calls _PyLong_New which always sets exception in cases it returns NULL: https://github.com/python/cpython/blob/6405feecda6a5d5dd7a4240eb3054a2676ed29b1/Objects/longobject.c#L203

Reply:
Can you provide a reproducible example for that exact case (so that `PyLong_FromUnsignedLongLong` does return `NULL`)?

We are just in the very last phase before the release and don't have much bandwidth. Since this does fix the memory leaks it should get "in". There is just not much time to do additional testing anymore. It should indeed be investigated and addressed better as mentioned in #201.

Reply:
Looking closer into how it is being called (https://github.com/ros2/rclpy/blob/869aa959f88a939260c8446f9c35ce301d71d23b/rclpy/src/rclpy/_rclpy.c#L2313-L2331) I think I am fine with keeping the else block (assuming the call sets an error) and return `NULL` which matches the logic of the caller. The same happens already when the `entity_type` is of invalid type.

Reply:
(sorry, used wrong account above)

This is harder to reproduce. From looking in Python code I can see this can happen only when Python runs out of memory (malloc fails). It will call PyErr_NoMemory() and return NULL. This means Python will raise MemoryError exception when control gets back to python code.

Reply:
I kept the else block (2031c485858ae86f1fdb6045f87cf20cad5f9b4d) after looking closer how it is being used and addressed a missing decref in the calling code (3b993a8d139a56fc83ab36e402e5ce2a2eaf316e) as well as eliminated the unreachable `return NULL` (6ca1439f9e346ea5db3bfc568e5e405d36784967). I hope that is fine for now...